arrows/vxl/average_frames.cxx

```diff
@@ -0,0 +1,740 @@
+bool online_frame_averager<PixType>
+::has_resolution_changed( const vil_image_view< PixType >& input )
+{
+  return ( input.ni() != this->last_average_.ni() ||
+           input.nj() != this->last_average_.nj() ||
+           input.nplanes() != this->last_average_.nplanes() );
+}
+
+// Helper function to allocate a completely new image, and cast the input image
+// to whatever specified type the output image is, scaling by some factor if set
+// and rounding if enabled, in one pass.
+template <typename inT, typename outT>
+void copy_cast_and_scale( const vil_image_view< inT >& input,
+                          vil_image_view< outT >& output,
+                          bool rounding_enabled,
+                          const inT scale_factor )
+{
+  // Just deep copy if pixel formats equivalent and there is no scale factor
+  if( vil_pixel_format_of(inT()) == vil_pixel_format_of(outT())
+      && scale_factor == 1.0 )
+  {
+    output.deep_copy( input );
+    return;
+  }
+
+  // Determine if any rounding would even be beneficial based on source types
+  if( rounding_enabled )
+  {
+    rounding_enabled = !std::numeric_limits<inT>::is_integer &&
+                       std::numeric_limits<outT>::is_integer;
+  }
+
+  // Resize, cast and copy
+  unsigned ni = input.ni(), nj = input.nj(), np = input.nplanes();
+  output.set_size( ni, nj, np );
+
+  // s* = source properties, d* = dest
+  std::ptrdiff_t sistep=input.istep(), sjstep=input.jstep(), spstep=input.planestep();
+  std::ptrdiff_t distep=output.istep(), djstep=output.jstep(), dpstep=output.planestep();
+
+  // Iterate through image and perform cast and any required ops. We could
+  // shorten this code using vil_transform and functors (binding arguments),
+  // but just to ensure they get inlined...
+  if( scale_factor == 1.0 && !rounding_enabled )
+  {
+    vil_convert_cast( input, output );
+  }
+  else if( !rounding_enabled )
+  {
+    const inT* splane = input.top_left_ptr();
+    outT* dplane = output.top_left_ptr();
+    for (unsigned p=0;p<np;++p,splane += spstep,dplane += dpstep)
+    {
+      const inT* srow = splane;
+      outT* drow = dplane;
+      for (unsigned j=0;j<nj;++j,srow += sjstep,drow += djstep)
+      {
+        const inT* spixel = srow;
+        outT* dpixel = drow;
+        for (unsigned i=0;i<ni;++i,spixel+=sistep,dpixel+=distep)
+        {
+          *dpixel = static_cast<outT>(*spixel * scale_factor);
+        }
+      }
+    }
+  }
+  else
+  {
+    const inT* splane = input.top_left_ptr();
+    outT* dplane = output.top_left_ptr();
+    for (unsigned p=0;p<np;++p,splane += spstep,dplane += dpstep)
+    {
+      const inT* srow = splane;
+      outT* drow = dplane;
+      for (unsigned j=0;j<nj;++j,srow += sjstep,drow += djstep)
+      {
+        const inT* spixel = srow;
+        outT* dpixel = drow;
+        for (unsigned i=0;i<ni;++i,spixel+=sistep,dpixel+=distep)
+        {
+          *dpixel = static_cast<outT>(*spixel * scale_factor + 0.5);
```

Comment:
There is a prior check in this function which ensures the output is integral in this case. This looks like it's circumventing the fact that `static_cast` truncates rather than rounds. I've replaced this whole thing with `vil_transform_*` things in #1184 .